File: examples/export-pages-and-assets/frameworks/playwright/nodejs/images.mjs
// Exports a page's HTML and linked assets (CSS, JS, images) to a local directory
// by intercepting network responses as the page loads.
//
// Install: npm install playwright-core
// Run:     node images.mjs

import fs from 'fs';
import path from 'path';
import { chromium } from 'playwright-core';

const TOKEN = 'YOUR_API_TOKEN_HERE';

const browser = await chromium.connectOverCDP(
  `wss://production-sfo.browserless.io?token=${TOKEN}`
);

try {
  const context = browser.contexts()[0];
  const page = await context.newPage();
  const assetPromises = [];

  page.on('response', (response) => {
    const type = response.request().resourceType();
    if (['stylesheet', 'script', 'image', 'font'].includes(type)) {
      assetPromises.push(
        response.body()
          .then((buf) => ({ url: response.url(), buf }))
          .catch(() => null)
      );
    }
  });

  await page.goto('https://example.com', { waitUntil: 'networkidle' });

  const html = await page.content();
  const assets = (await Promise.all(assetPromises)).filter(Boolean);

  fs.mkdirSync('page', { recursive: true });
  fs.writeFileSync('page/index.html', html);
  console.log('Saved page/index.html');

  for (const [i, { url, buf }] of assets.entries()) {
    const ext = path.extname(new URL(url).pathname) || '';
    const filename = `asset-${i}${ext}`;
    fs.writeFileSync(`page/${filename}`, buf);
    console.log(`Saved page/${filename}`);
  }
} finally {
  await browser.close();
}

File: examples/export-pages-and-assets/frameworks/playwright/python/images.py
# Exports a page's HTML and linked assets (CSS, JS, images) to a local directory
# by intercepting network responses as the page loads.
#
# Install: pip install playwright
# Run:     python images.py

import os
from urllib.parse import urlparse
from playwright.sync_api import sync_playwright

TOKEN = 'YOUR_API_TOKEN_HERE'
WS_ENDPOINT = f'wss://production-sfo.browserless.io?token={TOKEN}'

assets = []

def handle_response(response):
    if response.request.resource_type in ('stylesheet', 'script', 'image', 'font'):
        try:
            assets.append({'url': response.url, 'buf': response.body()})
        except Exception:
            pass

with sync_playwright() as playwright:
    browser = playwright.chromium.connect_over_cdp(WS_ENDPOINT)
    try:
        context = browser.contexts[0]
        page = context.new_page()
        page.on('response', handle_response)
        page.goto('https://example.com')
        page.wait_for_load_state('networkidle')

        html = page.content()

        os.makedirs('page', exist_ok=True)
        with open('page/index.html', 'w', encoding='utf-8') as f:
            f.write(html)
        print('Saved page/index.html')

        for i, asset in enumerate(assets):
            ext = os.path.splitext(urlparse(asset['url']).path)[1]
            filename = f'page/asset-{i}{ext}'
            with open(filename, 'wb') as f:
                f.write(asset['buf'])
            print(f'Saved {filename}')
    finally:
        browser.close()
